tests/RegistryTests.cpp

```diff
@@ -2971,3 +2971,244 @@ TEST_CASE("BasicRegistryTests::cotaskmem_array-bytes", "[registry]]")
     }
 }
 #endif // #if defined(__WIL_OBJBASE_H_)
+
+#if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
+TEST_CASE("BasicRegistryTests::enumerate_values", "[registry]]")
+{
+    const auto deleteHr = HRESULT_FROM_WIN32(::RegDeleteTreeW(HKEY_CURRENT_USER, testSubkey));
+    if (deleteHr != HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND))
+    {
+        REQUIRE_SUCCEEDED(deleteHr);
+    }
+
+    SECTION("enumerate_values with no values")
+    {
+        const wil::unique_hkey hkey{wil::reg::create_unique_key(HKEY_CURRENT_USER, testSubkey)};
+        const auto value_enum = wil::reg::enumerate_values(hkey.get());
+
+        REQUIRE(value_enum.begin() == value_enum.end());
+    }
+
+    SECTION("enumerate_values with one value - manual iterator usage")
+    {
+        wil::unique_hkey hkey{wil::reg::create_unique_key(HKEY_CURRENT_USER, testSubkey, wil::reg::key_access::readwrite)};
+        wil::reg::set_value(hkey.get(), L"first_value", 0);
+
+        const auto value_enum = wil::reg::enumerate_values(hkey.get());
+        REQUIRE(value_enum.begin() != value_enum.end());
+        auto value_iterator = value_enum.begin();
+        REQUIRE(std::wstring(*value_iterator) == L"first_value");
+        ++value_iterator;
+        REQUIRE(value_iterator == value_enum.end());
+
+        value_iterator = value_enum.begin();
+        REQUIRE(std::wstring(*value_iterator) == L"first_value");
+        value_iterator += 1;
+        REQUIRE(value_iterator == value_enum.end());
+
+        const auto uniquevalue_enum = wil::reg::enumerate_values(wistd::move(hkey));
+        REQUIRE(uniquevalue_enum.begin() != uniquevalue_enum.end());
+        value_iterator = uniquevalue_enum.begin();
+        REQUIRE(std::wstring(*value_iterator) == L"first_value");
+        ++value_iterator;
+        REQUIRE(value_iterator == uniquevalue_enum.end());
+
+        value_iterator = uniquevalue_enum.begin();
+        REQUIRE(std::wstring(*value_iterator) == L"first_value");
+        value_iterator += 1;
+        REQUIRE(value_iterator == uniquevalue_enum.end());
+    }
+
+    SECTION("enumerate_values with one value - std::for_each usage")
+    {
+        wil::unique_hkey hkey{wil::reg::create_unique_key(HKEY_CURRENT_USER, testSubkey, wil::reg::key_access::readwrite)};
+        wil::reg::set_value(hkey.get(), L"first_value", 0);
+
+        uint32_t count = 0;
+        auto enumerator = wil::reg::enumerate_values(hkey.get());

```

Comment:
consider if `wil::make_range` should be followed for this name. e.g. `wil::reg::make_value_range`, `wil::reg::make_key_range`.

I'm not sure how much thought went into "make range" so not sure how important that is to follow as precedent.

Reply:
Interesting. I didn't know wil had a make_range. Will next look at integrating that with this enumerator

Reply:
IMO we should follow the `std::filesystem::directory_iterator` pattern here where the iterator _itself_ exposes `begin()` and `end()` functions. E.g.

```c++
for (auto&& entry : std::directory_iterator(path)) { /* ... */ }
```

And then of course, a default-constructed iterator can be used as the `end` iterator

```c++
std::for_each(std::filesystem::directory_iterator(path), std::filesystem::directory_iterator, callback);
```

Reply:
I can rename this to be called "iterator" .. but to be honest, I don't understand a model where someone creates a "iterator" who *creates* other iterator types by exposing begin() and end(). That's completely confusing and goes against any iterator model I've seen in the STL.

Today it looks like this:
    for (const auto& value_data : wil::reg::enumerate_values(hkey))

but I can rename it to look like the filesystem name:
    for (const auto& value_data : wil::reg::value_iterator(hkey))

(same for wil::reg::key_iterator).

Duncan, Chris - thoughts?

Reply:
Or back to Chris's point: avoid 'iterator' nomenclature completely and embrace 'range'.

I'm perfectly happy to just call it:
    for (const auto& value_data : wil::reg::make_value_range(hkey))

(though that doesn't really follow "make_range" semantics)

or something like:

    for (const auto& value_data : wil::reg::make_value_range(reg_view_details::reg_view{ hkey })
    
... which can be moved out of the reg_view_details namespace of course.

Reply:
> That's completely confusing and goes against any iterator model I've seen in the STL.

The STL _is the one that set this precedent_ (the example I provide is _from_ the STL). Of course, not _all_ iterators in the STL behave this way as it wouldn't really make sense for container iterator types such as `std::vector<T>::iterator` or adapter iterator types such as `std::back_inserter<T>` to behave this way, and some iterator types that _could_ behave that way seem like they haven't been updated, such as `std::regex_iterator<T>` which predates the `std::filesystem` APIs.

> avoid 'iterator' nomenclature completely and embrace 'range'.

Ranges are built on top of iterators. In fact, the _definition_ of the concept `std::ranges::range` is that iterators can be made from the type. However, if we were to _truly_ embrace ranges, we'd call it a `view`, not a `range` (so something like `wil::reg::keys_view` and maybe `wil::reg::views::keys` - or if we really want to get crazy - `wil::reg::ranges::keys_view` and `wil::reg::ranges::views::keys`).

One benefit that using an iterator type here has is that it's easier to invoke "legacy" functions that expect iterators. E.g. compare the following two examples:

```c++
do_something(wil::reg::keys_iterator(path), wil::reg::keys_iterator());
```

vs

```c++
auto range = wil::reg::keys_view(path);
do_something(range.begin(), range.end());
```

Using an iterator type saves the local from needing to be created.

That said, I don't personally have too strong of opinions. `wil::reg::keys_iterator`, `wil::reg::make_keys_range`, and `wil::reg::keys_view` all sound good enough to me.

Reply:
Oh gosh, sorry - I must have replied entirely too late to see that of course, your example was from std::filesystem.

I'll look at their directory_iterator interface and usage.

Reply:
OK - I'm pushing changes that I think strongly aligns with this model. Thoughts?
this us of wil::reg::key_iterator always works perfectly with STL algorithms. (I have unit tests for std::for_each and std::count)

     for (const auto& key_data : wil::make_range(wil::reg::key_iterator{hkey}, wil::reg::key_iterator{}))
     {
         key_data.name; // the std::wstring of the enumerated key
     }

     for (const auto& key_data : wil::reg::enumerate_keys(hkey))
     {
         key_data.name; // the std::wstring of the enumerated key
     }

Reply:
I saw this and first thought was why wil:: and not some std type but wistd:: namespace

Reply:
Now documented [here](https://github.com/microsoft/wil/wiki/Registry-Helpers#enumeration) on the wiki